src/middleware/requireChangemakerPermission.ts

```diff
@@ -0,0 +1,38 @@
+import { Response, NextFunction } from 'express';
+import { InputValidationError, UnauthorizedError } from '../errors';
+import { isAuthContext, isId } from '../types';
+import type { Permission } from '../types';
+import type { Request } from 'express';
+
+const requireChangemakerPermission =
```

Comment:
This is an authorization middleware that checks whether the currently authenticated user has the given permission on the changemaker in the URL parameters. It is distinct from the handler that creates or deletes such permissions that this authorization handler checks, but happens to be used by the route leading to those handlers. It is also distinct from our authentication-checking handler but authentication is also checked here.

I'm trying to think if this could use a better name but perhaps not. It is in the `middleware` directory and not the `handlers` directory. I think it would be worth adding a comment on the function, though, basically my first paragraph, to remind our future selves.

Reply:
I'm not sure I understand, but to clarify the intention: this is express middleware (both in form and function) so `middleware` is the appropriate location.

The handler requires that the calling path have a `changemakerId` parameter, as well as an auth setting.  This is self documented IMHO because the function is simple and those checks are made.  It can be used on any path that has a `changemakerId` URL parameter.

Reply:
I meant to leave this comment here instead of in the other thread:

I think it's overstating it to say the handlers and middleware here are self-documenting. I think it becomes clearer after a long hard look, but I also think given the circular reference it may be worth a comment.

Reply:
But technically, yeah, you're right, it's a simple function, just read it carefully three or four times and you're good to go : )

Reply:
@bickelj I hear you -- given that point I think that it would be appropriate to add a docblock to all three "auth" middlewares to make it explicit what their intention is.